### PR TITLE
chore: optimise config backend revisionId calls

### PR DESCRIPTION
**Pull request by cisse21**
# Description

PrizePicks CFD

- Reduce the number of config backend API calls being made for a single revisionId
- `VACUUM FULL` notifier table if the size exceeds 5GB

## Linear Ticket

Fixes PIPE-1583

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/internal/loadfiles/loadfiles.go [hunk 1]
```diff
@@ -369,20 +374,20 @@ func (lf *LoadFileGenerator) createFromStaging(ctx context.Context, job *model.U
...
+	revisionIDMap[job.Warehouse.Destination.RevisionID] = job.Warehouse.Destination
+	revisionIds := lo.Uniq(
+		lo.Map(
+			lo.Filter(job.StagingFiles, func(file *model.StagingFile, _ int) bool {
+				return file.DestinationRevisionID != job.Warehouse.Destination.RevisionID
+			}), func(file *model.StagingFile, _ int) string {
+				return file.DestinationRevisionID
+			}))
+	for _, revisionId := range revisionIds {
+		destination, err := lf.ControlPlaneClient.DestinationHistory(ctx, revisionId)
```
Comment (Sidddddarth) [inline warehouse/internal/loadfiles/loadfiles.go:386]: Just adding a map check looks simpler than nested `lo.xxx`
```
func (lf *LoadFileGenerator) destinationRevisionIDMap(ctx context.Context, job *model.UploadJob) (revisionIDMap map[string]backendconfig.DestinationT, err error) {
	revisionIDMap = make(map[string]backendconfig.DestinationT)

	// TODO: ensure DestinationRevisionID is populated
	for _, file := range job.StagingFiles {
		revisionID := file.DestinationRevisionID
		// No need to make config backend api call for the current config
		if revisionID == job.Warehouse.Destination.RevisionID {
			revisionIDMap[revisionID] = job.Warehouse.Destination
			continue
		}
		// No need to make config backend api call for the same revision ID
		if _, ok := revisionIDMap[revisionID]; ok {
			continue
		}
		destination, err := lf.ControlPlaneClient.DestinationHistory(ctx, revisionID)
		if err != nil {
			return nil, err
		}
		revisionIDMap[revisionID] = destination
	}
	return
}
```

File: services/notifier/repo.go [hunk 2]
```diff
@@ -303,6 +306,19 @@ func (n *repo) deleteByBatchID(
...
+
+	var sizeEstimate int64
+	if err := n.db.QueryRowContext(
+		ctx,
+		fmt.Sprintf(`SELECT pg_table_size(oid) from pg_class where relname='%s';`, notifierTableName),
+	).Scan(&sizeEstimate); err != nil {
+		return fmt.Errorf("size estimate for notifierTable failed with: %w", err)
+	}
+	if sizeEstimate > config.GetInt64("Notifier.vacuumThresholdBytes", 5*bytesize.GB) {
+		if _, err := n.db.ExecContext(ctx, fmt.Sprintf(`vacuum full analyze %q;`, notifierTableName)); err != nil {
```
Comment (lvrach) [inline services/notifier/repo.go:318]: Did the full vacuuming helped here?

Comment (cisse21) [reply services/notifier/repo.go:318]: Yes it did. When the size got high the DB queries got slower affecting the throughput, so added the vacuum command to automate it

**Review by Sidddddarth — APPROVED** [review]

File: services/notifier/repo.go (hunk 2 above)
Comment (achettyiitr) [reply services/notifier/repo.go:318]: Should we do it from a master instead of a slave once all the track batch is completed?

Comment (cisse21) [reply services/notifier/repo.go:318]: Let's merge this and take that up if this causes any issues?

**Review by achettyiitr — APPROVED** [review]

**Pull request merged**